tests/node_test.cc

```diff
@@ -2,133 +2,250 @@
+
+        const int message_value = 0;
+
+        // Different type, never called
+        subscriber.Subscribe<TestMessage<10086>>(channel_subid, [](const auto&) {
+            // Never enters.
+            std::abort();
+        });
+
+        // Different subid, never called
+        subscriber.Subscribe<TestMessageType>(10010, [](const auto&) {
+            // Never enters.
+            std::abort();
+        });
+
+        subscriber.Subscribe<TestMessageType>(
+            channel_subid,
+            [message_value, cv, received](const std::shared_ptr<TestMessageType>& msg) {
+                EXPECT_EQ(msg->value_, message_value);
+                received->store(true);
+                cv->notify_all();
+            });
+
+        // duplicate subscription, first one wins
+        subscriber.Subscribe<TestMessageType>(channel_subid, [&](const auto&) {
+            // Never enters.
+            std::abort();
         });
-    }
 
-    std::this_thread::sleep_for(std::chrono::milliseconds(200));
-    EXPECT_EQ(unblocked.load(), 0);
-    std::this_thread::sleep_for(std::chrono::seconds(1));
-    ASSERT_EQ(unblocked.load(), kThreadNum);
-    for (auto&& thread : threads) {
-        thread.join();
+        publisher.Publish(channel_subid, std::make_shared<TestMessageType>(message_value));
+
+        std::unique_lock cv_lock(cv_mutex);
+        cv->wait(cv_lock, [received]() { return received->load(); });
     }
-}
 
-template<int idx>
-struct TestMessage : public CRMessage<TestMessage<idx>> {};
+    // Publishing while nobody subscribing.
+    publisher.Publish(channel_subid, std::make_shared<TestMessageType>(1));
+}
 
-TEST(NodeTest, Basic) {
-    CRNode                node(1);
-    constexpr std::size_t kNumOfTopics     = 4;
+TEST(NodeTest, MultipleChannels) {
+    constexpr std::size_t kNumOfMsgTypes   = 4;
     constexpr std::size_t kNumOfSubChannel = 4;
 
-    std::array<std::array<int, kNumOfSubChannel>, kNumOfTopics> received{};
+    CRNode publisher;
+    CRNode subscriber;
+    auto   runner = std::make_unique<JobRunner>(JobRunner::Config{});
+    subscriber.SetRunner(runner.get());
+
+    struct Received {
+        std::array<std::array<std::atomic<int>, kNumOfSubChannel>, kNumOfMsgTypes> data_{};
+    };
+
+    std::mutex cv_mutex;
+    auto       cv = std::make_shared<std::condition_variable>();
+
+    auto received = std::make_shared<Received>();
 
     for (std::size_t channel_subid = 0; channel_subid < kNumOfSubChannel; ++channel_subid) {
-        node.Subscribe<TestMessage<0>>(
+        subscriber.Subscribe<TestMessage<0>>(
             channel_subid,
-            [&received, channel_subid](const std::shared_ptr<TestMessage<0>>&) { ++received[0][channel_subid]; });
-        node.Subscribe<TestMessage<1>>(
+            [received, channel_subid, cv](const std::shared_ptr<TestMessage<0>>& msg) {
+                received->data_[0][channel_subid].store(msg->value_);
```

Comment:
If you really need atomic inside predictor, there's probably something wrong. Shared data for cv need to be gated by mutex, and the existence of atomic means someone else may forcefully change it during execution without acquiring the mutex first.